lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1993,7 +2036,188 @@ LogicalResult MemOp::canonicalize(MemOp op, PatternRewriter &rewriter) {
     }
   }
   rewriter.eraseOp(op);
-  return success();
+  return true;
+}
+
+// If memory has no write ports, eliminate it.
+bool foldReadOrWriteOnlyMemory(PatternRewriter &rewriter, MemOp op) {
+  bool isRead = false, isWritten = false;
+  for (unsigned i = 0; i < op.getNumResults(); ++i) {
+    switch (op.getPortKind(i)) {
+    case MemOp::PortKind::Read:
+      isRead = true;
+      if (isWritten)
+        return false;
+      continue;
+    case MemOp::PortKind::Write:
+      isWritten = true;
```

Comment:
How does this compose with unused ports ? So, if a memory has a read and write port but the write port is unused, will read only pattern be identified for such a memory ?

Reply:
Read-write ports have a single enable flag which disables both ends. I think what you're hinting at requires the inspection of the `wmode` flag, which is not done in this PR.

Reply:
I meant, Memory with one read port, one write port. If the write port is unused, the memory should be inferred as read-only. So, I was thinking, does it make sense to check `isPortUnused` before setting `isWritten`.